lib/Reconciler.lua

```diff
@@ -1,542 +1,361 @@
-]]
-local function applyRef(ref, newRbx)
-	if ref == nil then
-		return
+local function DEBUG_warn(...)
+	if DEBUG_LOGS then
+		warn(...)
 	end
+end
 
-	if type(ref) == "table" then
-		ref.current = newRbx
-	else
-		ref(newRbx)
+local function DEBUG_print(...)
+	if DEBUG_LOGS then
+		print(...)
 	end
 end
 
-local Reconciler = {}
-
-Reconciler._singleEventManager = SingleEventManager.new()
+local processTreeTasksSync
 
---[[
-	Is this element backed by a Roblox instance directly?
-]]
-local function isPrimitiveElement(element)
-	if type(element) ~= "table" then
-		return false
+local function DEBUG_showTask(task)
+	if typeof(task) == "function" then
+		return "<Task function>"
+	elseif typeof(task) == "table" then
+		return ("<Task %q>"):format(task.type)
+	else
+		return "<INVALID TASK>"
 	end
+end
 
-	return type(element.component) == "string"
+local function getGuid()
+	return HttpService:GenerateGUID(false)
 end
 
---[[
-	Is this element defined by a pure function?
-]]
-local function isFunctionalElement(element)
-	if type(element) ~= "table" then
-		return false
-	end
+local function scheduleTask(tree, task)
+	DEBUG_print("Scheduling task", DEBUG_showTask(task))
 
-	return type(element.component) == "function"
-end
+	tree.tasks[#tree.tasks + 1] = task
 
---[[
-	Is this element defined by a component class?
-]]
-local function isStatefulElement(element)
-	if type(element) ~= "table" then
-		return false
+	if not ASYNC_SCHEDULER and not tree.tasksRunning then
+		processTreeTasksSync(tree)
 	end
-
-	return type(element.component) == "table"
 end
 
---[[
-	Destroy the given Roact instance, all of its descendants, and associated
-	Roblox instances owned by the components.
-]]
-function Reconciler.unmount(instanceHandle)
-	local element = instanceHandle._element
+local function taskMountNode(details)
+	local element = details.element
+	local key = details.key
+	local parentNode = details.parentNode
+	local parentRbx = details.parentRbx
+	local isTreeRoot = details.isTreeRoot
+
+	assert(Type.is(element, Type.Element))
+	assert(typeof(key) == "string")
+	assert(Type.is(parentNode, Type.Node) or typeof(parentNode) == "nil")
+	assert(typeof(parentRbx) == "Instance" or typeof(parentRbx) == "nil")
+	assert(typeof(isTreeRoot) == "boolean")
+
+	return function(tree)
+		local node = {
+			[Type] = Type.Node,
+			children = {},
+			element = element,
+		}
 
-	if isPrimitiveElement(element) then
-		-- We're destroying a Roblox Instance-based object
+		if typeof(element.component) == "string" then
+			local rbx = Instance.new(element.component)
+			rbx.Name = key
 
-		-- Kill refs before we make changes, since any mutations past this point
-		-- aren't relevant to components.
-		applyRef(element.props[Core.Ref], nil)
+			node.rbx = rbx
+
+			for prop, value in pairs(element.props) do
+				if prop == Core.Children then
+					for childKey, childElement in pairs(value) do
+						node.children[childKey] = MountingNode
+
+						scheduleTask(tree, taskMountNode({
```

Comment:
It's a constructor!

Yeah, if we had static types with unions this could basically go away and it would be awesome. I've had that in TypeScript and it's soooo cool.